Feat 249 atomic

fmoessbauer opened the pull request:
This PR introduces a clean atomic concept in DASH. It provides an easy to use interface for atomic operations similar to `std::atomic<T>` for elements stored in DASH containers. The STL atomic concept cannot be ported 1:1 as remote accesses require copying (amongst others).

**The interface is as follows:**

- Phantom type `dash::Atomic<T>` which can (only) be placed in DASH containers.
- Specialisation of `dash::GlobRef<Atomic<T>>` which provides interface of `std::Atomic` and proxies it to atomic DART calls. Methods which cannot be performed atomically are disabled / deleted.
- Function-based interface `dash::atomic::` (works only for atomic types)
- DART interface for atomic operations which internally uses the corresponding MPI operations.
- Static assertions to check weather atomic operations on the underlying type are supported

**Currently neither implemented nor specified:**

- memory model information
- memory model traits
- handling of local accesses e.g. `array.lbegin().load()`

**Mixing atomic and non-atomic accesses**

As the constructor of `dash::GlobRef<T>` takes a DART gptr which is not strongly-typed, atomic accesses to non-atomic elements (and vice versa) are possible. In some cases this might be beneficial, as atomic operations are expensive compared to non-atomic ones.
However between sections with non-atomic and atomic operations there has to be a **barrier and a memory flush**. Otherwise the behavior is undefined. Furthermore we do not recommend using this "hack" at all.

An example of this is `dash::Shared` which provides both non-atomic (`shared.get()`) and atomic (`shared.atomic.get()`) access.

**Container Example:**
```c++
dash::Array<dash::Atomic<int>> array(100);
// supported as Atomic<value_t>(value_t T) is available
dash::fill(array.begin(), array.end(), 0);

if(dash::myid() == 0){
  array[10].store(5);
  }
dash::barrier();
array[10].add(1);
// postcondition:
// array[10] == dash::size() + 5
```

The following does not work:
```c++
dash::Array<dash::Atomic<int>> array(100);
array.local[10].load();               // not allowed
dash::atomic::load(array.local[10]);  // not allowed
dash::atomic::load(array.lbegin());   // not allowed
```

fmoessbauer committed:
made dash::Atomic trivially copyable by using dart team id instead of
reference to dash team [skip ci]

fmoessbauer committed:
fixed unit test for array of atomics

devreal committed:
Add sanity checks for DART communication functions

fmoessbauer committed:
add barrier after generation of atomics to fix race condition

fmoessbauer committed:
fixed and unified interface of dash::Atomic

fmoessbauer committed:
fixed dash::Atomic().set()

fmoessbauer committed:
added functional interface to atomically modify values of global objects

fuchsto reviewed (changes requested):
@dash-project/developers Do not use wrapper functions for atomic operations that hide the atomic type!

See my comment here:

https://github.com/dash-project/dash/issues/249#issuecomment-275514703

fmoessbauer committed:
Introduced new interface for DASH atomic semantic
- old `dash::Atomic` becomes `dash::AtomicAddress`
- new version of `dash::Atomic` is phantom type
- provide algorithms to change GlobRef<Atomic<>>
- specialize GlobRef<Atomic> to provide atomic interface
- name atomic interface according to std::atomic
- documentation

fmoessbauer commented:
As discussed yesterday, I finally managed to change the interface in a way that follows as close as possible the C++11 standard as well as `std::atomic`.

This introduces the following changes:
- old `dash::Atomic` becomes `dash::AtomicAddress`
- new version of `dash::Atomic` is phantom type
- provide algorithms to change `GlobRef<Atomic<>>`
- specialize `GlobRef<Atomic>` to provide atomic interface
- name atomic interface according to std::atomic

Please review it very carefully.

fmoessbauer committed:
Moved implementation of GlobRef<Atomic> to specialized class,
documentation, cleanup

fmoessbauer committed:
fixed interface of += and -= operators

fmoessbauer committed:
fixed indentation of GlobRef code

fmoessbauer committed:
make GlobRef atomic operations const, as the GlobRef does not own the
atomic values

fmoessbauer committed:
implemented atomic exchange operation

fmoessbauer committed:
added interface for compare_exchange and cleanup

fmoessbauer committed:
stability fixes in atomic test

fmoessbauer committed:
explicitly flush segment in atomic.store

devreal commented on dash/include/dash/Atomic.h:
```diff
@@ -34,58 +39,56 @@ class Atomic
+    DASH_LOG_DEBUG_VAR("AtomicAddress.set()", value);
+    DASH_LOG_TRACE_VAR("AtomicAddress.set",   _gptr);
+    *(pointer(_gptr)) = value;
+    dart_flush_all(_gptr);
```
Two things: 

1. Is the flush really required? The assignment should use `dart_put_blocking`, which issues a flush at the end. 
2. More importantly: This operation is not atomic and might thus conflict with other atomic operations executed on this memory. I suggest using `dart_accumulate`, which is guaranteed to be atomic (implemented in terms of `MPI_Accumulate`.

devreal commented on dash/include/dash/Atomic.h:
```diff
@@ -34,58 +39,56 @@ class Atomic
+    DASH_LOG_DEBUG("AtomicAddress.get()");
+    DASH_LOG_TRACE_VAR("AtomicAddress.get", _gptr);
     DASH_ASSERT(!DART_GPTR_ISNULL(_gptr));
     reference ref(_gptr);
```
See my comment on `set()`. Better use `dart_fetch_and_op` with `DART_OP_NONE`. Alternatively, we might consider implementing a counterpart of `MPI_Get_accumulate`, i.e., `dart_get_accumulate`.

devreal commented on dash/include/dash/Atomic.h:
```diff
@@ -202,19 +204,288 @@ class Atomic
+    /// value to store in the atomic object if it is as expected
+    ValueType desired)
+  {
+    // TODO;
```
I guess this requires something like `dart_compare_and_swap()`, will add an implementation of it today.

devreal commented on dash/include/dash/GlobRef.h:
```diff
@@ -335,24 +338,260 @@ class GlobRef
     return member<MEMTYPE>(offs);
   }
-
+  
```
grrr, why all those trailing whitespaces?

devreal commented on dash/include/dash/GlobRef.h:
```diff
@@ -335,24 +338,260 @@ class GlobRef
+  
+};
+
+
```
I suggest moving this specialization to `Atomic.h` as non-atomic use does not require it to compile successfully.

devreal reviewed (changes requested):
Great work! Some suggestions for improvements inline. I will take care of implementing `dart_compare_and_swap` today.

devreal committed:
Implement dart_compare_and_swap and dash::AtomicAddress::compare_exchange; Add type check for dash::AtomicAddress

fmoessbauer committed:
restructured dash atomic implementations, types and algorithms

fmoessbauer committed:
avoid explicit flushes and use atomic dart operations

fmoessbauer commented:
@devreal Thanks for your comments. I did some major restructuring of the atomic types / classes and algorithms. Now the interface is clean and can be loaded by including `dash/Atomic.h`.

And now to the technical part: I changed the implementation according to your comments, but unfortunately now the atomics do not work anymore. For me it seems that the dart versions of MPI accumulate functions lose some of the atomicity guarantees. So please have a look at the changes. In the meantime we could use commit 347cbdb which works perfectly.

fuchsto commented:
@fmoessbauer: Good, please wait for my review before further implementation.

fuchsto committed:
Merge branch 'development' into feat-249-atomic

devreal commented:
@fmoessbauer Thanks for restructuring the code, much better now. You mean that it does not work because of the Segfault? I'm not sure whether this is related to `dart_fetch_and_op`, as the segfault happens in in a flush:

```
#0  0x00007ffff5d64428 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007ffff5d6602a in __GI_abort () at abort.c:89
#2  0x00007ffff5da67ea in __libc_message (do_abort=do_abort@entry=1, fmt=fmt@entry=0x7ffff5ebd8a2 "*** %s ***: %s terminated\n") at ../sysdeps/posix/libc_fatal.c:175
#3  0x00007ffff5e4756c in __GI___fortify_fail (msg=<optimized out>, msg@entry=0x7ffff5ebd884 "stack smashing detected") at fortify_fail.c:37
#4  0x00007ffff5e47510 in __stack_chk_fail () at stack_chk_fail.c:28
#5  0x00007ffff08721ac in ?? () from /usr/lib/openmpi/lib/openmpi/mca_btl_vader.so
#6  0x00007ffff50091ea in opal_progress () from /usr/lib/libopen-pal.so.13
#7  0x00007fffeaf70895 in ?? () from /usr/lib/openmpi/lib/openmpi/mca_osc_pt2pt.so
#8  0x00007fffeaf71667 in ompi_osc_pt2pt_flush_all () from /usr/lib/openmpi/lib/openmpi/mca_osc_pt2pt.so
#9  0x00007ffff6d7cf04 in PMPI_Win_flush_all () from /usr/lib/libmpi.so.12
#10 0x000000000071fca8 in dart_flush_all (gptr=...) at /home/joseph/src/dash/dash/dart-impl/mpi/src/dart_communication.c:1108
#11 0x0000000000606bad in dash::GlobMem<dash::Atomic<int>, dash::allocator::CollectiveAllocator<dash::Atomic<int> > >::flush_all (this=<optimized out>)
    at /home/joseph/src/dash/dash/dash/include/dash/GlobMem.h:420
#12 dash::Array<dash::Atomic<int>, long, dash::BlockPattern<1, (dash::MemArrange)1, long> >::barrier (this=0x7fffffffd160, this=0x7fffffffd160)
    at /home/joseph/src/dash/dash/dash/include/dash/Array.h:1082
#13 dash::Array<dash::Atomic<int>, long, dash::BlockPattern<1, (dash::MemArrange)1, long> >::deallocate (this=this@entry=0x7fffffffd160)
    at /home/joseph/src/dash/dash/dash/include/dash/Array.h:1174
#14 0x0000000000600dae in dash::Array<dash::Atomic<int>, long, dash::BlockPattern<1, (dash::MemArrange)1, long> >::~Array (this=0x7fffffffd160, 
    __in_chrg=<optimized out>) at /home/joseph/src/dash/dash/dash/include/dash/Array.h:819
#15 AtomicTest_AtomicInterface_Test::TestBody (this=<optimized out>) at /home/joseph/src/dash/dash/dash/test/AtomicTest.cc:205
#16 0x00000000006e80eb in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)
    ()
#17 0x00000000006e16ef in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) ()
#18 0x00000000006c4e2c in testing::Test::Run() ()
#19 0x00000000006c57c4 in testing::TestInfo::Run() ()
#20 0x00000000006c5eb7 in testing::TestCase::Run() ()
#21 0x00000000006ccffe in testing::internal::UnitTestImpl::RunAllTests() ()
#22 0x00000000006e96c3 in bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) ()
#23 0x00000000006e2531 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) ()
#24 0x00000000006cba9a in testing::UnitTest::Run() ()
#25 0x00000000004d8a92 in RUN_ALL_TESTS () at /home/joseph/src/dash/dash/build/gtest/src/GTestExternal/googletest/include/gtest/gtest.h:2233
#26 main (argc=1, argv=0x7fffffffd938) at /home/joseph/src/dash/dash/dash/test/main.cc:70
```

Is that the Segfault you see as well? This error appeared using OpenMPI 1.10.2. I will test other versions and MPICH as well.

devreal commented:
It seems that the `fetch_and_op` operation corrupts the stack in some way in the test suite. However, I am not able to reproduce this with a small example so it's painful to debug. I'll keep digging through it...

devreal committed:
Slightly extended debug messages

devreal committed:
Fix strange segfaults/stack corruptions after issueing atomic operations. Apparently, a full flush is required (at some point) to ensure correctness and avoid UB, so flush on each access.

fmoessbauer commented:
@devreal Thanks for putting time into this. Yesterday I talked to @rkowalewski and @fuchsto about the flushes. We need them, as we have to ensure remote completion. However I guess they are not necessary for pure read operations?

devreal commented:
That's what I realized today, I was assuming that atomic operations are performed on the spot. I guess I was too puzzled by the random errors the missing flushes caused, all MPI implementations are terrible in these cases. Inside `get()`, a call to `dart_flush_local` should be sufficient. Not sure whether that makes any difference though.

devreal committed:
dart_flush_local should be sufficient in AtomicAddress::get()

devreal committed:
compare_and_swap is only valid on integral types

devreal committed:
Remove unnecessary team parameter from atomic operations

fuchsto committed:
Merge branch 'development' into feat-249-atomic

fuchsto committed:
Merge branch 'development' into feat-249-atomic

fuchsto committed:
Merge branch 'development' into feat-249-atomic

fuchsto commented on dash/include/dash/atomic/AtomicAddress.h:
```diff
@@ -0,0 +1,270 @@
+class GlobRef;
+
+template<typename ValueType>
+class AtomicAddress 
```
For consistency, this class template should be named `GlobAtomicRef`. It acts as extension of `GlobRef` by atomic operations, like `GlobAsyncRef` does for async operations.
Leading to my second point: This type should satisfy (and extend) the `GlobRef` concept. I think it already does, but it should be checked and documented.

The name `AtomicAddress` is misleading / incorrect in any case:
An address is not atomic, it's just a (compound) number.
That also is to say: the template class does not provide operations on an **address**, and this would be nonsensical. So let's not use this term here.

Let me illustrate in simplified pseudo code:

```c++
Address a(something);
Address b = a.sub(3); // <-- looks like pointer arithmetic?
```

fuchsto reviewed (changes requested):
The memory model traits are still missing, but I wouldn't care for now. The changes definitely are an improvement compared to the current implementation in `development`.

The memory model concept will be addressed in a separate issue.

fuchsto committed:
Merge branch 'development' into feat-249-atomic

fmoessbauer committed:
Finalized implementation of `dash::Atomic` by removing `AtomicAddress`.
This also includes many bugfixes, unit tests and cleanup

fuchsto committed:
Merge branch 'development' into feat-249-atomic

fuchsto commented on dash/include/dash/atomic/GlobAtomicRef.h:
```diff
@@ -0,0 +1,367 @@
+class GlobRef<dash::Atomic<T>>
+{
+  static_assert(dash::dart_datatype<T>::value != DART_TYPE_UNDEFINED,
+         "dash::AtomicAddress only valid on integral and floating point types");
```
Should be `dash::GlobRef<dash::Atomic<T>>` in static assertion message

fuchsto commented on dash/include/dash/atomic/GlobAtomicRef.h:
```diff
@@ -0,0 +1,367 @@
+   */
+  void store(const T & value) const
+  {
+    DASH_LOG_DEBUG_VAR("AtomicAddress.set()", value);
```
Log messages have scope "AtomicAddress"

fuchsto committed:
Merge branch 'development' into feat-249-atomic

fuchsto committed:
Merge branch 'development' into feat-249-atomic

fmoessbauer committed:
fixed trace messages and inline comments for dash::GlobRef<Atomic>

fmoessbauer committed:
added documentation of DASH's atomic interface in userguide

fuchsto commented:
@fmoessbauer Are all comments by @devreal addressed in your recent changes?

fuchsto committed:
Merge branch 'development' into feat-249-atomic
